### DISCUSSION TITLE
Linux demo 纯CPU下的耗时分析

### DISCUSSION THREAD

**Discussion by zhusonger** [Tencent/libpag#1716] [Q&A] [answered]
大佬好, 目前想在Linux服务器上部署该功能, 碰到一些优化方面的疑问，望解答，谢谢

相似问题 #1099 #1125 

# 场景
运行Linux demo, 分别统计flush和readPixels耗时

flush方法: 一个0.0x毫秒,2个几十毫秒, 偶尔出现几百毫秒
> ---currentFrame:0, flushStatus:1, timeCost:2537.839000 ms
---currentFrame:1, flushStatus:0, timeCost:0.019000 ms
---currentFrame:2, flushStatus:1, timeCost:2.980000 ms
---currentFrame:3, flushStatus:1, timeCost:15.936000 ms
---currentFrame:12, flushStatus:1, timeCost:373.597000 ms
---currentFrame:31, flushStatus:0, timeCost:0.022000 ms
---currentFrame:32, flushStatus:1, timeCost:34.722000 ms
---currentFrame:33, flushStatus:1, timeCost:183.007000 ms
---currentFrame:34, flushStatus:0, timeCost:0.019000 ms
---currentFrame:35, flushStatus:1, timeCost:21.831000 ms
---currentFrame:36, flushStatus:1, timeCost:27.537000 ms

readPixels方法, 基本稳定在几十毫秒
> ---currentFrame:0, flushStatus:1, timeCost:46.464000 ms
---currentFrame:1, flushStatus:0, timeCost:36.969000 ms
---currentFrame:2, flushStatus:1, timeCost:60.816000 ms
---currentFrame:3, flushStatus:1, timeCost:51.333000 ms
---currentFrame:4, flushStatus:0, timeCost:47.597000 ms
---currentFrame:5, flushStatus:1, timeCost:50.016000 ms
---currentFrame:6, flushStatus:1, timeCost:60.495000 ms
---currentFrame:7, flushStatus:0, timeCost:37.109000 ms
---currentFrame:8, flushStatus:1, timeCost:55.287000 ms
---currentFrame:9, flushStatus:1, timeCost:40.641000 ms
---currentFrame:10, flushStatus:0, timeCost:48.710000 ms
---currentFrame:11, flushStatus:1, timeCost:50.198000 ms
---currentFrame:12, flushStatus:1, timeCost:45.642000 ms

# 电脑配置
* 内存: 16G
* CPU: Intel® Core™ i5-7500 CPU @ 3.40GHz × 4

# 提问
## Q3: demo里面没有涉及到GPU的优化, 但是执行demo的时候, 发现GPU会有占用, 是哪里占用的呢? readPixels方法吗?

**Comment by domchen** [comment, marked as answer]
1.flush的耗时不是通过优化调用方式，是通过优化素材。你之所以好几帧波动特别大，是因为那几帧的画面复杂度特别高。通过官网的素材优化指南文档对素材优化就行。一次只有一个线程可以调用flush，内部是有加锁的，你如果并发起来调用其他线程都会阻塞住，没有任何意义。但是linux有个其他的优化思路，就是多进程渲染，你可以把一个素材分时间切片渲染，运行多个程序只生成其中的一部分，比如你是要生成视频片段，保持视频输出的格式完全一致，最后一个ffmpeg指令合并起来就行，不涉及任何编解码就很快。但是这个方式有个别的问题，首帧建立渲染环境的开销是很高的，当你切片过多的时候就会开始到临界点时间无法再加快了。

2.readPixels在Linux的CPU模式上没有任何优化办法。你最终想拿的就是CPU数据，怎么都得读取出来。

3.现在的Linux Demo并不是实现了CPU渲染，是靠SwiftShader模拟的GPU，SwiftShader就是demo项目里带的那几个EGL的动态库。结论是它没有任何可能执行GPU指令。另外如果你的Linux设备是带了GPU的，你可以直接替换自带的那几个EGL的动态库为你系统上的真的调用GPU的EGL动态库，重新链接一下你就可以真的用GPU渲染了。

**Reply by zhusonger** [reply]
好的 谢谢您的解答
